### PR TITLE
changefeedccl: improve Kafka batching logic

### PR DESCRIPTION
**Pull request by HonoreDB**
Our Kafka client batches by default and is configurable to batch
even more, which can lead to errors from the server if the batch
is too large.

This PR fixes this in two ways. If the batch size is one, we're
in an irrecoverable state without manual intervention, so we
no longer mark that specific error as retryable in that case.
Also, we set the sink to not batch until its first successful
flush. That ensures that even if we retry, we eventually make
progress.

Addresses https://github.com/cockroachdb/cockroach/issues/80313

Release note (enterprise change): Improved handling of Kafka message size too large error.

### CODE REVIEW

File: pkg/ccl/changefeedccl/BUILD.bazel
Comment (miretskiy) [inline pkg/ccl/changefeedccl/BUILD.bazel:18]: Did you forget to add this file?

File: pkg/ccl/changefeedccl/sink_kafka.go
Comment (miretskiy) [inline pkg/ccl/changefeedccl/sink_kafka.go:261]: I don't see the implementation of this method, but I can guess... And i'm worried.
I suppose it's perfectly possible that a blip happens when feed runs w/out batching, and you fail to emit the message.  We probably don't want to mark feed failed.

I thought that at some point we will be relying on a "sane-er" retry behavior, where coordinator notices that it has retried many times, but the water mark does not advance.

**Comment by shermanCRL** [comment]
Nice. For my understanding, if we have already “transitioned to” batching, which I understand to mean “batching only after the first successful flush”, and we subsequently encounter this error -- do we then “reset” back to the “don’t batch until successful flush”?

Overall, I think this sounds great -- instead of reducing batch size, we just go to batch size 1 until it works.

**Comment by HonoreDB** [comment]
> For my understanding, if we have already “transitioned to” batching, which I understand to mean “batching only after the first successful flush”, and we subsequently encounter this error -- do we then “reset” back to the “don’t batch until successful flush”?
Yup! If we encounter the error in the batching state, we mark it as retryable, so the changefeed gets restarted from its last checkpoint and restarts in the no batching state.

**Comment by samiskin** [comment]
Let's say events A, B, C, D, and E all fit into a batch, D is the one message that when batched results in "message too large"

1. We put them all into a batch, the batch fails, they all show up with message too large.  
2. We retry the changefeed with the new config.  A, B, and C all succeed 
3. We Flush, the flush succeeds, and we move back to the batching sink. 
4. D now appears in the batching sink, more events are added to the batch, the batch is flushed, and we see the message too large error.
5. The changefeed restarts, but it hasn't yet placed a new checkpoint, so it restarts back before A, and we return to step 2.

We then keep looping through steps 2-5

Does the above sound possible?

File: pkg/ccl/changefeedccl/BUILD.bazel
Comment (HonoreDB) [reply pkg/ccl/changefeedccl/BUILD.bazel:18]: Yup, pushed it.

File: pkg/ccl/changefeedccl/sink_kafka.go
Comment (HonoreDB) [reply pkg/ccl/changefeedccl/sink_kafka.go:261]: Implementation is lower down in the same file and it should address your concern (but take a look!) It only marks fatal when it's the specific Kafka message too large error and batching is disabled.

**Comment by HonoreDB** [comment]
>The changefeed restarts, but it hasn't yet placed a new checkpoint
That's a great point, I'm treating flushing and checkpointing as one atomic event but they're certainly not. I'm not sure whether it's possible, I kind of want to say no because the same worker is responsible for sending the checkpoint and reporting the subsequent error? I'll see if I can prove or disprove that.

**Review by miretskiy — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84668)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @HonoreDB, and @samiskin)

___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 293 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84668#-N7QbDxDCNqNbQyeqrgi:-N7QbDxDCNqNbQyeqrgj:bp63n9l) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b15910d285f46ba9628e5d9923544f898c3d4e3/pkg/ccl/changefeedccl/sink_kafka.go#L293)):*
> ```Go
> 	}
> 
> 	return s.maybeMarkFatal(s.topics.Each(func(topic string) error {
> ```

I wonder if a simpler, and an identical (I think) mechanism of ensuring things are flushed
is to just
```
return errors.CombineErrors(
   s.maybeMarkFatal(....),
   s.Flush()
)
```



_Code quote:_
```Go
return s.maybeMarkFatal(s.topics.Each(func(topic string) error {
```
___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 347 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84668#-N7QbRCW23avl6GFLgi8:-N7QbRCW23avl6GFLgi9:b-99aclh) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b15910d285f46ba9628e5d9923544f898c3d4e3/pkg/ccl/changefeedccl/sink_kafka.go#L347)):*
> ```Go
> 	defer func() {
> 
> 	}()
> ```

return 42? or... what's the meaning of this function :)
___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 442 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84668#-N7QcUGi4gPJihpKbYNw:-N7QcUGj1vpjK79cCe3h:b-db1to8) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b15910d285f46ba9628e5d9923544f898c3d4e3/pkg/ccl/changefeedccl/sink_kafka.go#L442)):*
> ```Go
> 			s.mu.flushCh <- struct{}{}
> 			s.mu.flushCh = nil
> 		}
> ```

I think @samiskin brought up some good points.  I'm going to type up my suggestions -- but please treat them as such -- I'm not sure that this will work...

First, I think it's important that from the API perspective, the current guarantees remain: Namely, if you EmitMessage, then this message is emitted or an error is returned.  Batching, and error handling should be an implementation detail.

Second, I think we should be conservative as to what kind of error messages get "special" treatment -- just like you have now.  Namely, unless we *know* that the reason is message too large, we shouldn't internally retry, but should fall back to retrying the whole feed.

Now, as to the mechanics of how to accomplish this, I would suggest that when we encounter message too large error, drain the queue, then retry.  I think doing that is not too crazy:

I would add 2 things to s.mu struct:

```
 retry []*sarama.ProducerMessage
 readyToResume sync.Cond
```

With those 2 things added, the worker loop enqueues messages into []retry if there was an error.

I think sarama guarantees that these acks will arrive in the right order; but maybe we should be
paranoid and maybe queue those resolved messages at the end of []retry (or just have a single
retryResolved message -- because there should only be 1).

Now, when we enter retry mode, we ought to block any new messages from coming in,
until we sort things out; luckily all messages enter the system through emitMessage, which
acquires the lock; so, it's easy to do something like this, before actually emitting the message:

```
  if len(s.mu.retry) > 0 {
     s.mu.readyToResume.Wait()
  }
```

That is, we'll block until retry becomes empty again (and we'll need to signal when that happens).


After we enqueue message for retry:
```
   if s.mu.inflight == 0 && len(s.mu.retry) > 0 {
       return doRetry()
   }
```

doRetry or whatever will need to exit workerLoop since we'll need to recreate producer -- (likely producer needs to be under mu).
Critically, any pending emitMessages (both resolved message and real messages) are now blocked until
we sort things out.

I guess it would make sense to use current "stats" to see what was the largest message; what was the size of the batch that we think put us over the edge and act accordingly.

**Review by miretskiy — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84668)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @HonoreDB, and @samiskin)

___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 442 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84668#-N7QcUGi4gPJihpKbYNw:-N7QiPW8Eu4fYd0JRgP0:b623hry) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b15910d285f46ba9628e5d9923544f898c3d4e3/pkg/ccl/changefeedccl/sink_kafka.go#L442)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I think @samiskin brought up some good points.  I'm going to type up my suggestions -- but please treat them as such -- I'm not sure that this will work...

First, I think it's important that from the API perspective, the current guarantees remain: Namely, if you EmitMessage, then this message is emitted or an error is returned.  Batching, and error handling should be an implementation detail.

Second, I think we should be conservative as to what kind of error messages get "special" treatment -- just like you have now.  Namely, unless we *know* that the reason is message too large, we shouldn't internally retry, but should fall back to retrying the whole feed.

Now, as to the mechanics of how to accomplish this, I would suggest that when we encounter message too large error, drain the queue, then retry.  I think doing that is not too crazy:

I would add 2 things to s.mu struct:

```
 retry []*sarama.ProducerMessage
 readyToResume sync.Cond
```

With those 2 things added, the worker loop enqueues messages into []retry if there was an error.

I think sarama guarantees that these acks will arrive in the right order; but maybe we should be
paranoid and maybe queue those resolved messages at the end of []retry (or just have a single
retryResolved message -- because there should only be 1).

Now, when we enter retry mode, we ought to block any new messages from coming in,
until we sort things out; luckily all messages enter the system through emitMessage, which
acquires the lock; so, it's easy to do something like this, before actually emitting the message:

```
  if len(s.mu.retry) > 0 {
     s.mu.readyToResume.Wait()
  }
```

That is, we'll block until retry becomes empty again (and we'll need to signal when that happens).


After we enqueue message for retry:
```
   if s.mu.inflight == 0 && len(s.mu.retry) > 0 {
       return doRetry()
   }
```

doRetry or whatever will need to exit workerLoop since we'll need to recreate producer -- (likely producer needs to be under mu).
Critically, any pending emitMessages (both resolved message and real messages) are now blocked until
we sort things out.

I guess it would make sense to use current "stats" to see what was the largest message; what was the size of the batch that we think put us over the edge and act accordingly.
</blockquote></details>

Forget what I wrote about sync.Cond -- use a channel that's always closed, unless we are in the retry loop.
Need the channel because, unfortunately, sync.Cond is not context aware.

**Review by HonoreDB — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84668)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @miretskiy, and @samiskin)

___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 293 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84668#-N7QbDxDCNqNbQyeqrgi:-N7RZHEr3YpxkyEBUuYt:buz0wt8) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b15910d285f46ba9628e5d9923544f898c3d4e3/pkg/ccl/changefeedccl/sink_kafka.go#L293)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I wonder if a simpler, and an identical (I think) mechanism of ensuring things are flushed
is to just
```
return errors.CombineErrors(
   s.maybeMarkFatal(....),
   s.Flush()
)
```

</blockquote></details>

I don't think we want the sink to force a flush before emitting resolved timestamps--I like that it's nonblocking unless something has gone very wrong already. 
___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 347 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84668#-N7QbRCW23avl6GFLgi8:-N7RZhiQ07j5p_kQb6NQ:b-oclxxh) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b15910d285f46ba9628e5d9923544f898c3d4e3/pkg/ccl/changefeedccl/sink_kafka.go#L347)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

return 42? or... what's the meaning of this function :)
</blockquote></details>

...hope I didn't mean to put something important there.
___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 442 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84668#-N7QcUGi4gPJihpKbYNw:-N7R_2RyEft-NbvbJXhK:b2546ng) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b15910d285f46ba9628e5d9923544f898c3d4e3/pkg/ccl/changefeedccl/sink_kafka.go#L442)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Forget what I wrote about sync.Cond -- use a channel that's always closed, unless we are in the retry loop.
Need the channel because, unfortunately, sync.Cond is not context aware.
</blockquote></details>

Post-caffeine I am now able to understand this comment--since the producer errors contain the original failed message, we can retry each of them with batching disabled. That seems worth trying! If and when you approve this PR I'll merge it as-is, but whether you approve it now or not I'll see if I can get your better retry to work. I might be able to do the blocking part more simply within the multisink idiom by calling asapSink.Flush() at the top of EmitRow() (was that what your other comment was saying?).

**Review by miretskiy — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84668)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @HonoreDB, @miretskiy, and @samiskin)

___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 293 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84668#-N7QbDxDCNqNbQyeqrgi:-N7S1yZzA8SzA9HXDkId:b-ez6txc) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b15910d285f46ba9628e5d9923544f898c3d4e3/pkg/ccl/changefeedccl/sink_kafka.go#L293)):*
<details><summary><i>Previously, HonoreDB (Aaron Zinger) wrote…</i></summary><blockquote>

I don't think we want the sink to force a flush before emitting resolved timestamps--I like that it's nonblocking unless something has gone very wrong already. 
</blockquote></details>

But is it enough reason? We emit those messages very infrequently (hopefully); and those messages
can only be emitted after we checkpoint (which is also a slow operation).... Is the added complexity worth it?

**Comment by miretskiy** [comment]
> I might be able to do the blocking part more simply within the multisink idiom by calling asapSink.Flush()

I guess what I want to know is: is it worth having asap sink at all?
Also, if this drain/retry works, we don't have to retry 1 message at a time...  We could for example, create configuration with smaller batch size.. Maybe try it few times.. 1024 messages -> 512 -> 256 -> give up, switch to 1 message...
And if that fails.. well, it fails.

**Review by samiskin — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84668)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @miretskiy, and @samiskin)

___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 293 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84668#-N7QbDxDCNqNbQyeqrgi:-N7SAz3gDlcMs4WqLkZ3:b-qpsohe) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b15910d285f46ba9628e5d9923544f898c3d4e3/pkg/ccl/changefeedccl/sink_kafka.go#L293)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

But is it enough reason? We emit those messages very infrequently (hopefully); and those messages
can only be emitted after we checkpoint (which is also a slow operation).... Is the added complexity worth it?
</blockquote></details>

Since this is only done on the frontier we'd only be flushing batched resolved messages anyway so it shouldn't be that large of a flush.

**Review by HonoreDB — COMMENTED** [review]
If the drain/retry scheme works then sure, probably no reason I can't just create a new sink (or new producer) on the fly rather than initializing two sinks at startup, I'll see what's simplest.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84668#-:-N7SGk90BSYKbRwtZalM:bxg29r8)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @miretskiy, and @samiskin)

___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 293 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84668#-N7QbDxDCNqNbQyeqrgi:-N7SFIG50qFSp10tBBVk:b-kdk9v3) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b15910d285f46ba9628e5d9923544f898c3d4e3/pkg/ccl/changefeedccl/sink_kafka.go#L293)):*
<details><summary><i>Previously, samiskin (Shiranka Miskin) wrote…</i></summary><blockquote>

Since this is only done on the frontier we'd only be flushing batched resolved messages anyway so it shouldn't be that large of a flush.
</blockquote></details>

To be clear, there's no added complexity in this method, just an added comment to note that this method no longer batches (and marking fatal if a resolved message gets a MessageSizeTooLarge error, which should definitely always be fatal). It's true I could get back the old behavior by adding a call to Flush, but the old behavior was wrong anyway, it just gets fixed here as a side effect.

**Comment by shermanCRL** [comment]
Sounds like we’ve decided to focus on https://github.com/cockroachdb/cockroach/pull/86138 ?

**Pull request closed without merging**
